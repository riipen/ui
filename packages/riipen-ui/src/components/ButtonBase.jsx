import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";
import _JSXStyle from "styled-jsx/style";

const ButtonBase = ({
  ariaLabel,
  children,
  classes,
  type,
  variant,
  ...other
}) => {
  const className = clsx("button", variant, classes);
  return (
    <>
      <button
        aria-label={ariaLabel}
        className={className}
        type={type}
        {...other}
      >
        {children}
      </button>
      <style jsx>{`
        .button {
          cursor: pointer;
        }

        .button:focus {
          outline: 5px auto -webkit-focus-ring-color;
        }

        .default {
          background-color: transparent;
          border: none;
          text-decoration: none;
        }
      `}</style>
    </>
  );
};

ButtonBase.propTypes = {
  // external

  /**
   * The aria-label to provide to the button
   */
  ariaLabel: PropTypes.string.isRequired,

  /**
   * The children to render within the button
   */
  children: PropTypes.node.isRequired,

  /**
   * List of additional classes to apply to this component.
   */
  classes: PropTypes.oneOfType([PropTypes.array, PropTypes.string]),

  /**
   * The variant for the button. Default clears normal html button styles.
   */
  variant: PropTypes.oneOf(["default"]),

  /**
   * Whether traditional button styles (border, background) should be cleared
   */
  type: PropTypes.oneOf(["button", "submit", "reset"])
};

ButtonBase.defaultProps = {
  classes: [],
  variant: "default",
  type: "button"
};

ButtonBase.displayName = "ButtonBase";

export default ButtonBase;
