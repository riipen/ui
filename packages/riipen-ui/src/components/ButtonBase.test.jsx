import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import ButtonBase from "./ButtonBase";

describe("ButtonBase component", () => {
  let props;
  let children;

  beforeEach(() => {
    children = <div>A child</div>;
    props = {
      ariaLabel: "click to be cool"
    };
  });

  it("renders without errors", () => {
    let error;

    try {
      mount(<ButtonBase {...props}>{children}</ButtonBase>);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<ButtonBase {...props}>{children}</ButtonBase>);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const wrapper = mount(<ButtonBase {...props}>{children}</ButtonBase>);

      expect(wrapper.find("ButtonBase").props().classes).toEqual([]);
      expect(wrapper.find("ButtonBase").props().variant).toEqual("default");
      expect(wrapper.find("ButtonBase").props().type).toEqual("button");
    });
  });

  describe("children prop", () => {
    it("displays given children", () => {
      const wrapper = mount(<ButtonBase {...props}>{children}</ButtonBase>);

      expect(wrapper.containsMatchingElement(children)).toBeTruthy();
    });
  });

  describe("classes prop", () => {
    it("sets custom class", () => {
      props.classes = "newClass";

      const wrapper = mount(<ButtonBase {...props}>{children}</ButtonBase>);

      expect(
        wrapper
          .find("ButtonBase")
          .childAt(0)
          .hasClass(props.classes)
      ).toEqual(true);
    });
  });

  describe("variant prop", () => {
    it("applies variant class", () => {
      props.variant = "default";

      const wrapper = mount(<ButtonBase {...props}>{children}</ButtonBase>);

      expect(wrapper.find("ButtonBase").props().variant).toEqual(props.variant);
      expect(
        wrapper
          .find("ButtonBase")
          .find("button")
          .hasClass(props.variant)
      ).toEqual(true);
    });
  });

  describe("type prop", () => {
    it("applies type to button", () => {
      const type = "submit";

      const wrapper = mount(
        <ButtonBase {...props} type={type}>
          {children}
        </ButtonBase>
      );

      expect(wrapper.find("ButtonBase").props().type).toEqual(type);
    });
  });

  describe("ariaLabel prop", () => {
    it("applies ariaLabel to button", () => {
      props.ariaLabel = "cool";

      const wrapper = mount(<ButtonBase {...props}>{children}</ButtonBase>);

      expect(wrapper.find("ButtonBase").props().ariaLabel).toEqual(
        props.ariaLabel
      );
      expect(wrapper.find("button[aria-label='cool']")).toHaveLength(1);
    });
  });
});
