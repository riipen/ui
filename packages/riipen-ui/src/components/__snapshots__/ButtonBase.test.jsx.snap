// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonBase component renders correct snapshot 1`] = `
<ButtonBase
  ariaLabel="click to be cool"
  classes={Array []}
  type="button"
  variant="default"
>
  <button
    aria-label="click to be cool"
    className="jsx-509404678 button default"
    type="button"
  >
    <div>
      A child
    </div>
  </button>
  <JSXStyle
    id="509404678"
  />
</ButtonBase>
`;
